pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseUtils.java

```diff
@@ -0,0 +1,226 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.blocks;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import org.apache.commons.collections.MapUtils;
+import org.apache.pinot.common.datatable.DataTable;
+import org.apache.pinot.common.exception.QueryException;
+import org.apache.pinot.common.request.context.ExpressionContext;
+import org.apache.pinot.common.response.ProcessingException;
+import org.apache.pinot.common.utils.DataSchema;
+import org.apache.pinot.core.common.datatable.DataTableBuilderUtils;
+import org.apache.pinot.core.operator.blocks.results.ExceptionResultsBlock;
+import org.apache.pinot.core.operator.blocks.results.MetadataResultsBlock;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.function.DistinctAggregationFunction;
+import org.apache.pinot.core.query.distinct.DistinctTable;
+import org.apache.pinot.core.query.request.context.QueryContext;
+import org.apache.pinot.core.query.request.context.utils.QueryContextUtils;
+
+
+public final class InstanceResponseUtils {
+  private InstanceResponseUtils() {
+    // do not instantiate.
+  }
+
+  public static InstanceResponseBlock getEmptyResponseBlock(DataSchema explainResultSchema) {
+    return new InstanceResponseBlock(new MetadataResultsBlock(explainResultSchema), null);
+  }
+
+  public static InstanceResponseBlock getExceptionBlock(ProcessingException exception) {
+    return new InstanceResponseBlock(new ExceptionResultsBlock(exception), null);
+  }
+
+  public static DataTable toDataTable(InstanceResponseBlock instanceResponseBlock) {
+    DataTable dataTable;
+    try {
+      if (instanceResponseBlock.getBaseResultsBlock() == null) {
+        dataTable = DataTableBuilderUtils.buildEmptyDataTable();
+      } else {
+         dataTable = instanceResponseBlock.getBaseResultsBlock().getDataTable(instanceResponseBlock.getQueryContext());
+      }
+      if (MapUtils.isNotEmpty(instanceResponseBlock.getExceptionMap())) {
+        for (Map.Entry<Integer, String> e : instanceResponseBlock.getExceptionMap().entrySet()) {
+          dataTable.addException(e.getKey(), e.getValue());
+        }
+      }
+      if (MapUtils.isNotEmpty(instanceResponseBlock.getInstanceResponseMetadata())) {
+        dataTable.getMetadata().putAll(instanceResponseBlock.getInstanceResponseMetadata());
+      }
+      return dataTable;
+    } catch (ProcessingException pe) {
+      return toDataTable(getExceptionBlock(pe));
+    } catch (Exception e) {
+      return toDataTable(getExceptionBlock(QueryException.UNKNOWN_ERROR));
+    }
+  }
+
+  public static Collection<Object[]> toRows(InstanceResponseBlock instanceResponseBlock) {
```

Comment:
Can we add a brief javadoc like separator to clearly indicate the functions that are added for the multi-stage engine optimization. Just so future readers have sufficient context on how this is used

Reply:
Actually the optimization is applicable to current engine as well if broker and server are running together I guess ?